eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Tally.hs

```diff
@@ -0,0 +1,99 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE EmptyDataDeriving #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedLists #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module Cardano.Ledger.Conway.Rules.Tally (
+  ConwayTALLY,
+  TallyEnv (..),
+  GovernanceProcedure (..),
+  TallyPredicateFailure,
+) where
+
+import Cardano.Ledger.BaseTypes (ShelleyBase)
+import Cardano.Ledger.Conway.Era (ConwayTALLY)
+import Cardano.Ledger.Conway.Governance (
+  ConwayTallyState (..),
+  GovernanceActionId (..),
+  GovernanceActionInfo (..),
+  GovernanceActionState (..),
+  Vote (..),
+  makeGovAction,
+ )
+import Cardano.Ledger.Core (Era (..))
+import Cardano.Ledger.Shelley.Tx (TxId (..))
+import Control.State.Transition.Extended (
+  STS (..),
+  TRC (..),
+  TransitionRule,
+  judgmentContext,
+ )
+import qualified Data.Map.Strict as Map
+import Data.Sequence (Seq, ViewL (..), viewl)
+import qualified Data.Sequence as Seq
+
+newtype TallyEnv era = TallyEnv (TxId (EraCrypto era))
+
+data GovernanceProcedure era
+  = VotingProcedure (Vote era)
+  | ProposalProcedure (GovernanceActionInfo era)
+
+data TallyPredicateFailure era
+  deriving (Eq, Show)
+
+instance Era era => STS (ConwayTALLY era) where
+  type State (ConwayTALLY era) = ConwayTallyState era
+  type Signal (ConwayTALLY era) = Seq (GovernanceProcedure era)
+  type Environment (ConwayTALLY era) = TallyEnv era
+  type BaseM (ConwayTALLY era) = ShelleyBase
+  type PredicateFailure (ConwayTALLY era) = TallyPredicateFailure era
+  type Event (ConwayTALLY era) = ()
+
+  initialRules = []
+
+  transitionRules = [tallyTransition]
+
+addVote ::
+  GovernanceActionId (EraCrypto era) ->
+  Vote era ->
+  ConwayTallyState era ->
+  ConwayTallyState era
+addVote gaid vote@Vote {..} (ConwayTallyState st) =
+  ConwayTallyState $
+    Map.update (pure . updateVote) gaid st
+  where
+    updateVote GovernanceActionState {..} =
+      GovernanceActionState
+        { gasVotes = Map.insert (voteRole, voteRoleKeyHash) vote gasVotes
```

Comment:
Somewhere we need to check that the voting roles are legitimate. e.g., that someone claiming to be a constitutional committee member really is such.

Reply:
Actually, the spec doesn't say anything about checking the keyhashes, but I need to add a check for whether the governance action being voted for actually exists.